scripts/docgen/generate-docs.js

```diff
@@ -0,0 +1,133 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const { exec } = require('child-process-promise');
+const { argv } = require('yargs');
+const fs = require('mz/fs');
+const path = require('path');
+const yaml = require('js-yaml');
+
+const repoPath = path.resolve(`${__dirname}/../..`);
+const docPath = path.resolve(`${__dirname}/html`);
+const sourceFile = argv.source
+  ? path.resolve(argv.source)
+  : `${repoPath}/packages/firebase/index.d.ts`;
+
+const command = `${repoPath}/node_modules/.bin/typedoc ${sourceFile} \
+--out ${docPath} \
+--readme ${__dirname}/HOME_TEMP.md \
+--options ${__dirname}/typedoc.js \
+--theme ${__dirname}/theme`;
+
+console.log('Running command:\n', command);
+
+function moveFilesToRoot(subdir) {
+  return exec(`mv ${docPath}/${subdir}/* ${docPath}`)
+    .then(() => {
+      exec(`rmdir ${docPath}/${subdir}`);
+    })
+    .catch(e => console.error(e));
+}
+
+// Reformat links to match flat structure.
+function fixLinks(path) {
```

Comment:
I think we need more detailed comments to make the process easier to understand since there is a lot of customizations happening and they are very specific to our needs. 

For example, for each function and step I'd add to the comment why we do it as well.

Reply:
Broke out a lot of functionality into separate functions, commented each function and its call.